File: cpp/benchmarks/codebookmethods.cpp
/**
 * @file codebookmethods.cpp
 * @brief Nanobench performance benchmark for Codebook's own lookup/validation methods
 */

#define ANKERL_NANOBENCH_IMPLEMENT
#include <nanobench.h>

#include <dnv/VistaSDK.h>

namespace dnv::vista::sdk::benchmarks
{
    int run()
    {
        const auto& vis = VIS::instance();
        const auto& codebooks = vis.codebooks(VisVersion::v3_11a);

        const auto& quantity = codebooks[CodebookName::Quantity]; // 247 standard values, 1 group
        const auto& state = codebooks[CodebookName::State];       // 148 standard values, 59 groups
        const auto& position = codebooks[CodebookName::Position]; // 35 standard values, 12 groups

        ankerl::nanobench::Bench bench;
        bench.title("CodebookMethods").warmup(10000).minEpochIterations(5000000);

        bench.run("Quantity_hasStandardValue_hit", [&] {
            bool a = quantity.hasStandardValue("temperature");
            ankerl::nanobench::doNotOptimizeAway(a);
        });

        bench.run("Quantity_hasStandardValue_miss", [&] {
            bool a = quantity.hasStandardValue("not_a_real_quantity");
            ankerl::nanobench::doNotOptimizeAway(a);
        });

        bench.run("Quantity_createTag_standard", [&] {
            auto a = quantity.createTag("temperature");
            ankerl::nanobench::doNotOptimizeAway(a);
        });

        bench.run("Quantity_createTag_custom", [&] {
            auto a = quantity.createTag("custom_measurement");
            ankerl::nanobench::doNotOptimizeAway(a);
        });

        bench.run("State_hasGroup_hit", [&] {
            bool a = state.hasGroup("Running");
            ankerl::nanobench::doNotOptimizeAway(a);
        });

        bench.run("State_hasGroup_miss", [&] {
            bool a = state.hasGroup("NotARealGroup");
            ankerl::nanobench::doNotOptimizeAway(a);
        });

        bench.run("Position_validatePosition_simple", [&] {
            auto a = position.validatePosition("centre");
            ankerl::nanobench::doNotOptimizeAway(a);
        });

        bench.run("Position_validatePosition_composite", [&] {
            auto a = position.validatePosition("centre-starboard-2");
            ankerl::nanobench::doNotOptimizeAway(a);
        });

        return 0;
    }
} // namespace dnv::vista::sdk::benchmarks

int main()
{
    return dnv::vista::sdk::benchmarks::run();
}

File: bindings/rust/vista-sdk/benches/codebook_methods.rs
use criterion::{criterion_group, criterion_main, Criterion};
use vista_sdk::core::codebook_name::CodebookName;
use vista_sdk::core::vis::Vis;
use vista_sdk::core::vis_version::VisVersion;

fn bench(c: &mut Criterion) {
    let vis = Vis::instance();
    let codebooks = vis
        .codebooks(VisVersion::V3_11a)
        .expect("V3_11a should be valid");

    let quantity = &codebooks[CodebookName::Quantity];
    let state = &codebooks[CodebookName::State];
    let position = &codebooks[CodebookName::Position];

    c.bench_function("Quantity_hasStandardValue_hit", |b| {
        b.iter(|| quantity.has_standard_value("temperature"))
    });

    c.bench_function("Quantity_hasStandardValue_miss", |b| {
        b.iter(|| quantity.has_standard_value("not_a_real_quantity"))
    });

    c.bench_function("Quantity_createTag_standard", |b| {
        b.iter(|| quantity.create_tag("temperature"))
    });

    c.bench_function("Quantity_createTag_custom", |b| {
        b.iter(|| quantity.create_tag("custom_measurement"))
    });

    c.bench_function("State_hasGroup_hit", |b| {
        b.iter(|| state.has_group("Running"))
    });

    c.bench_function("State_hasGroup_miss", |b| {
        b.iter(|| state.has_group("NotARealGroup"))
    });

    c.bench_function("Position_validatePosition_simple", |b| {
        b.iter(|| position.validate_position("centre"))
    });

    c.bench_function("Position_validatePosition_composite", |b| {
        b.iter(|| position.validate_position("centre-starboard-2"))
    });
}

criterion_group!(benches, bench);
criterion_main!(benches);
